### PR TITLE
feat:  create component to block screen depending on config

### PR DESCRIPTION
**Pull request by gca-axelor**
RM#65925

### CODE REVIEW

**Comment by gca-axelor** [comment]
/*
 * Axelor Business Solutions
 *
 * Copyright (C) 2023 Axelor (<http://axelor.com>).
 *
 * This program is free software: you can redistribute it and/or  modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, {useCallback, useEffect} from 'react';
import {StyleSheet, View, Dimensions} from 'react-native';
import {
  Screen,
  SwitchCard,
  useConfig,
  useTheme,
  RightIconButton,
  Icon,
  useThemeColor,
  Text,
  BlockInteractionMessage,
  Button,
} from '@axelor/aos-mobile-ui';
import {clearMessage, uploadTranslations} from '../features/configSlice';
import {useDispatch, useSelector} from 'react-redux';
import {getTranslations, selectLanguage, useTranslator} from '../../i18n';
import {showToastMessage} from '../../utils';
import {
  disable,
  enable,
  useEffectOnline,
  useOnline,
} from '../../features/onlineSlice';
import {ApiProviderConfig} from '../../apiProviders/config';
import {useNavigation} from '@axelor/aos-mobile-core';

const SettingsScreen = ({children}) => {
  const Colors = useThemeColor();
  const I18n = useTranslator();
  const Theme = useTheme();
  const online = useOnline();
  const dispatch = useDispatch();
  const navigation = useNavigation();

  const {
    showFilter,
    hideVirtualKeyboard,
    setActivityIndicator,
    toggleFilterConfig,
    toggleVirtualKeyboardConfig,
    setShowSubtitles,
    showSubtitles,
  } = useConfig();
  const language = useSelector(selectLanguage);

  const {message} = useSelector(state => state.config);
  const {user} = useSelector(state => state.user);
  const {baseUrl} = useSelector(state => state.auth);

  const {setShowBlockInteractionMessage} = useConfig();

  useEffect(() => {
    if (message) {
      showToastMessage({type: 'success', position: 'bottom', text1: message});
      setActivityIndicator(false);
      dispatch(clearMessage());
    }
  }, [message, dispatch, setActivityIndicator]);

  const handleToggleConnection = useCallback(
    state => {
      if (!state) {
        dispatch(enable());
      } else {
        dispatch(disable());
      }
    },
    [dispatch],
  );

  useEffectOnline();

  const handleToggleColorBlind = useCallback(
    state => {
      if (state) {
        Theme.activateColorBlind();
      } else {
        Theme.desactivateColorBlind();
      }
    },
    [Theme],
  );

  const handleToggleSubtitles = useCallback(
    state => {
      setShowSubtitles(state);
    },
    [setShowSubtitles],
  );

  const handleSendTranslations = useCallback(() => {
    setActivityIndicator(true);
    const translations = getTranslations(language);
    dispatch(uploadTranslations({language, translations}));
  }, [dispatch, language, setActivityIndicator]);

  return (
    <Screen style={styles.screen}>
      <View style={styles.container}>
        <SwitchCard
          title={I18n.t('User_ShowFilter')}
          defaultValue={showFilter}
          onToggle={toggleFilterConfig}
        />
        <SwitchCard
          title={I18n.t('User_VirtualKeyboardConfig')}
          defaultValue={hideVirtualKeyboard}
          onToggle={toggleVirtualKeyboardConfig}
        />
        <SwitchCard
          title={I18n.t('User_ColorForColorBlind')}
          defaultValue={Theme.isColorBlind}
          onToggle={handleToggleColorBlind}
        />
        {ApiProviderConfig.allowConnectionBlock && (
          <SwitchCard
            title={I18n.t('User_BlockConnection')}
            defaultValue={!online.isEnabled}
            onToggle={handleToggleConnection}
          />
        )}
        <SwitchCard
          title={I18n.t('User_Show_Drawer_Subtitles')}
          defaultValue={showSubtitles}
          onToggle={handleToggleSubtitles}
        />
        {children}
        {user?.group?.code !== 'admins' ? null : (
          <RightIconButton
            icon={
              <Icon
                name="chevron-right"
                color={Colors.primaryColor.background}
              />
            }
            style={styles.RightIconButton}
            title={I18n.t('User_SendTranslations')}
            onPress={handleSendTranslations}
          />
        )}
      </View>
      <View style={styles.footerContainer}>
        <Text>{`${I18n.t('Base_ConnectedOn')}:`}</Text>
        <Text numberOfLines={1}>{baseUrl}</Text>
      </View>
      <BlockInteractionMessage navigation={navigation} />
      <Button
        onPress={() => setShowBlockInteractionMessage(true, 'teeeeeeeest')}
      />
    </Screen>
  );
};

const styles = StyleSheet.create({
  screen: {
    justifyContent: 'space-between',
  },
  container: {
    marginTop: 15,
    justifyContent: 'flex-start',
    alignItems: 'center',
    flex: 1,
  },
  RightIconButton: {
    width: Dimensions.get('window').width * 0.9,
    height: 40,
    marginLeft: 18,
    paddingRight: 50,
    paddingLeft: 10,
  },
  footerContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 10,
  },
});

export default SettingsScreen;

**Comment by gca-axelor** [comment]
<Button
        onPress={() =>
          setShowBlockInteractionMessage({
            show: true,
            message: 'teeeeeeeest',
            callback: navigation.popToTop,
            iconName: 'yin-yang',
          })
        }
      />

**Comment by gca-axelor** [comment]
<Button
        onPress={() =>
          setBlockInteractionConfig({
            visible: true,
            message: 'teeeeeeeest',
            callback: navigation.popToTop,
            iconName: 'yin-yang',
          })
        }
      />

File: packages/ui/src/components/molecules/BlockInteractionMessage/BlockInteractionMessage.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      <Card style={styles.container}>
+        <WarningCard
+          errorMessage={blockInteractionConfig.message}
+          style={styles.warningCard}
+        />
+        <Icon
+          name={blockInteractionConfig?.iconName}
+          touchable={true}
+          onPress={handleButton}
+        />
```
Comment (lme-axelor) [inline packages/ui/src/components/molecules/BlockInteractionMessage/BlockInteractionMessage.tsx:55]: I think we can do better than just one icon in the middle 😄 

As I already told you, I think we should enable people to give a list of actions as, depending on the context, we can have multiple actions possible on this pop-up : refresh the app, disconnect, try again, ... 

So my suggestion is to have three props in blockInteractionConfig :
- visible : to know if component should be visible
- message : to fill in WarningCard
- actions : a list of actions to display composed with : 
```
{
    iconName?: string
    tittle: string
    onPress: () => void
    color?: Color
}
```
and this list should be displayed with Button, or IconButton depending on presence of iconName, in a column style, button should take most of the width of the pop-up like it's done in `PopupApplicationNotConfigured`

File: packages/ui/src/config/ConfigContext.tsx [hunk 2]
```diff
@@ -142,6 +190,17 @@ const configReducer = (
...
 
 const actions = {
+  setBlockInteractionConfig: ({visible, message, iconName, callback}) => {
+    return {
+      type: actionTypes.setBlockInteractionConfig,
+      payload: {
+        visible: visible,
+        message: message,
+        iconName,
+        callback: callback,
```
Comment (lme-axelor) [inline packages/ui/src/config/ConfigContext.tsx:200]: That's funny how only iconName transmission is simplified 😆 
```
        visible,
        message,
        iconName,
        callback,
```

File: packages/ui/src/config/ConfigContext.tsx [hunk 3]
```diff
@@ -30,6 +30,23 @@ interface ConfigContextState {
   showActivityIndicator: boolean;
   showSubtitles: boolean;
   headerHeight: number;
+  blockInteractionConfig: {
+    visible: boolean;
+    message: string;
+    iconName: string;
+    callback: () => void;
+  };
```
Comment (lme-axelor) [inline packages/ui/src/config/ConfigContext.tsx:38]: I think you defined this type like three or four times, do not hesitate to create real interface and use this interface in the rest of the code 
```
 interface BlockInteractionConfig {
    visible: boolean;
    message: string;
    iconName: string;
    callback: () => void;
  };
```

**Pull request merged**